middleware/session/middleware.go

```diff
@@ -0,0 +1,188 @@
+package session
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/gofiber/fiber/v3"
+	"github.com/gofiber/fiber/v3/log"
+)
+
+// key for looking up session middleware in request context
+const key = 0
+
+// Session defines the session middleware configuration
+type MiddlewareConfig struct {
+	// Next defines a function to skip this middleware when returned true.
+	//
+	// Optional. Default: nil
+	Next func(c fiber.Ctx) bool
+
+	// Store defines the session store
+	//
+	// Required.
+	Store *Store
+
+	// ErrorHandler defines a function which is executed for errors
+	//
+	// Optional. Default: nil
+	ErrorHandler func(*fiber.Ctx, error)
+}
+
+type Middleware struct {
+	config     MiddlewareConfig
+	Session    *Session
+	ctx        *fiber.Ctx
+	hasChanged bool
+	mu         sync.RWMutex
+}
+
+// Middleware pool
+var middlewarePool = &sync.Pool{
+	New: func() any {
+		return &Middleware{}
+	},
+}
+
+// Session is a middleware to manage session state
+//
+// Session middleware manages common session state between requests.
+// This middleware is dependent on the session store, which is responsible for
+// storing the session data.
+func NewMiddleware(config MiddlewareConfig) fiber.Handler {
+	return func(c fiber.Ctx) error {
+		// Don't execute middleware if Next returns true
+		if config.Next != nil && config.Next(c) {
+			return c.Next()
+		}
+
+		// Get the session
+		session, err := config.Store.Get(c)
+		if err != nil {
+			return err
+		}
+
+		// get a middleware from the pool
+		m := acquireMiddleware()
+		m.config = config
+		m.Session = session
+		m.ctx = &c
+
+		// Store the middleware in the context
+		c.Locals(key, m)
+
+		// Continue stack
+		stackErr := c.Next()
+
+		// Save the session
+		// This is done after the response is sent to the client
+		// It allows us to modify the session data during the request
+		// Without having to worry about calling Save()
+		//
+		// It will also extend the session idle timeout automatically.
+		if err := session.Save(); err != nil {
+			if config.ErrorHandler != nil {
+				config.ErrorHandler(&c, err)
+			} else {
+				log.Errorf("session: %v", err)
+			}
+		}
+
+		// release the middleware back to the pool
+		releaseMiddleware(m)
+
+		return stackErr
+	}
+}
+
+var ErrTypeAssertionFailed = errors.New("failed to type-assert to *Middleware")
+
+// acquireMiddleware returns a new Middleware from the pool
+func acquireMiddleware() *Middleware {
+	middleware, ok := middlewarePool.Get().(*Middleware)
+	if !ok {
+		panic(fmt.Errorf("%w", ErrTypeAssertionFailed))
```

Comment:
Here you are generating unnecessary processing.
```suggestion
		panic(ErrTypeAssertionFailed.Error())
```